pkg/dataplane/http/context.go

```diff
@@ -308,54 +315,151 @@ func (c *context) GetItemsSync(getItemsInput *v3io.GetItemsInput) (*v3io.Respons
-	if err != nil {
-		return nil, err
-	}
+		getItemsResponse := struct {
+			Items            []map[string]map[string]interface{}
+			NextMarker       string
+			LastItemIncluded string
+		}{}
 
-	//validate getItems response to avoid infinite loop
-	if getItemsResponse.LastItemIncluded != "TRUE" && (getItemsResponse.NextMarker == "" || getItemsResponse.NextMarker == getItemsInput.Marker) {
-		errMsg := fmt.Sprintf("Invalid getItems response: lastItemIncluded=false and nextMarker='%s', "+
-			"startMarker='%s', probably due to object size bigger than 2M. Query is: %+v", getItemsResponse.NextMarker, getItemsInput.Marker, getItemsInput)
-		c.logger.Warn(errMsg)
-	}
+		// unmarshal the body into an ad hoc structure
+		err = json.Unmarshal(response.Body(), &getItemsResponse)
+		if err != nil {
+			return nil, err
+		}
 
-	getItemsOutput := v3io.GetItemsOutput{
-		NextMarker: getItemsResponse.NextMarker,
-		Last:       getItemsResponse.LastItemIncluded == "TRUE",
-	}
+		//validate getItems response to avoid infinite loop
+		if getItemsResponse.LastItemIncluded != "TRUE" && (getItemsResponse.NextMarker == "" || getItemsResponse.NextMarker == getItemsInput.Marker) {
+			errMsg := fmt.Sprintf("Invalid getItems response: lastItemIncluded=false and nextMarker='%s', "+
+				"startMarker='%s', probably due to object size bigger than 2M. Query is: %+v", getItemsResponse.NextMarker, getItemsInput.Marker, getItemsInput)
+			c.logger.Warn(errMsg)
+		}
+
+		getItemsOutput := v3io.GetItemsOutput{
+			NextMarker: getItemsResponse.NextMarker,
+			Last:       getItemsResponse.LastItemIncluded == "TRUE",
+		}
+
+		// iterate through the items and decode them
+		for _, typedItem := range getItemsResponse.Items {
 
-	// iterate through the items and decode them
-	for _, typedItem := range getItemsResponse.Items {
+			item, err := c.decodeTypedAttributes(typedItem)
+			if err != nil {
+				return nil, err
+			}
 
-		item, err := c.decodeTypedAttributes(typedItem)
+			getItemsOutput.Items = append(getItemsOutput.Items, item)
+		}
+		// attach the output to the response
+		response.Output = &getItemsOutput
+	} else {
+		r := bytes.NewReader(response.Body())
+		capnpSections := readAllCapnpMessages(r)
+		if len(capnpSections) < 2 {
+			return response, errors.Errorf("getItemsCapnp: Got only %v capnp sections. Expecting at least 2", len(capnpSections))
+		}
+		cookie := string(response.HeaderPeek("X-v3io-cookie"))
+		getItemsOutput := v3io.GetItemsOutput{
+			NextMarker: cookie,
+			Last:       len(cookie) == 0,
+		}
+		if len(capnpSections) < 2 {
+			return response, errors.Errorf("getItemsCapnp: Got only %v capnp sections. Expecting at least 2", len(capnpSections))
+		}
+
+		metadataPayload, err := vncapnp.ReadRootVnObjectItemsGetResponseMetadataPayload(capnpSections[len(capnpSections) - 1])
 		if err != nil {
-			return nil, err
+			return response, err
+		}
+		//Keys
+		attributeMap, err := metadataPayload.KeyMap()
+ 		if err != nil {
+ 			return response, err
```

Comment:
Please use `errors.Wrap(err, <reason>)` here - otherwise we'll have no idea what the underlying error is